pkg/kubernetes/container/v1alpha1/container.go

```diff
@@ -0,0 +1,194 @@
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package v1alpha1
+
+import (
+	"github.com/pkg/errors"
+	corev1 "k8s.io/api/core/v1"
+)
+
+type container struct {
+	corev1.Container // kubernetes container type
+}
+
+// OptionFunc is a typed function that abstracts anykind of operation
+// against the provided container instance
+//
+// This is the basic building block to create functional operations
+// against the container instance
+type OptionFunc func(*container)
+
+// Predicate abstracts conditional logic w.r.t the container instance
+//
+// NOTE:
+// Predicate is a functional approach versus traditional approach to mix
+// conditions such as *if-else* within blocks of business logic
+//
+// NOTE:
+// Predicate approach enables clear separation of conditionals from
+// imperatives i.e. actions that form the business logic
+type Predicate func(*container) (nameOrMsg string, ok bool)
+
+// predicateFailedError returns the provided predicate as an error
+func predicateFailedError(message string) error {
+	return errors.Errorf("predicatefailed: %s", message)
+}
+
+var (
+	validationFailedError = errors.New("container validation failed")
+)
+
+// asContainer transforms this container instance into corresponding kubernetes
+// container type
+func (c *container) asContainer() corev1.Container {
+	return corev1.Container{
+		Name:                     c.Name,
+		Image:                    c.Image,
+		Command:                  c.Command,
+		Args:                     c.Args,
+		WorkingDir:               c.WorkingDir,
+		Ports:                    c.Ports,
+		EnvFrom:                  c.EnvFrom,
+		Env:                      c.Env,
+		Resources:                c.Resources,
+		VolumeMounts:             c.VolumeMounts,
+		VolumeDevices:            c.VolumeDevices,
+		LivenessProbe:            c.LivenessProbe,
+		ReadinessProbe:           c.ReadinessProbe,
+		Lifecycle:                c.Lifecycle,
+		TerminationMessagePath:   c.TerminationMessagePath,
+		TerminationMessagePolicy: c.TerminationMessagePolicy,
+		ImagePullPolicy:          c.ImagePullPolicy,
+		SecurityContext:          c.SecurityContext,
+		Stdin:                    c.Stdin,
+		StdinOnce:                c.StdinOnce,
+		TTY:                      c.TTY,
+	}
+}
+
+// New returns a new kubernetes container
+func New(opts ...OptionFunc) corev1.Container {
+	c := &container{}
+	for _, o := range opts {
+		o(c)
+	}
+	return c.asContainer()
+}
+
+// builder provides utilities required to build a kubernetes container type
+type builder struct {
+	con    *container  // container instance
+	checks []Predicate // validations to be done while building the container instance
+	errors []error     // errors found while building the container instance
+}
+
+// Builder returns a new instance of builder
```

Comment:
`// Builder returns a new instance of builder`  --> `// Builder returns a new empty instance of builder`

Reply:
I am focusing more on what it provides versus what it contains. Need not be followed as a rule though.